README.md
# pindo-cli

![Tests](https://github.com/pindoio/pindo-cli/workflows/Tests/badge.svg)
![fury](https://badge.fury.io/py/pindo-cli.svg)
![pipy](https://pypip.in/d/pindo-cli/badge.png)

## Installation

Install from PyPi using [pip](http://www.pip-installer.org/en/latest), a package manager for Python.

`pip install pindo-cli`

Don't have pip installed? Try installing it, by running this from the command line:

`$ curl https://raw.github.com/pypa/pip/master/contrib/get-pip.py | python`

`python setup.py install`

You may need to run the above commands with `sudo`.

## Getting Started

Once you have installed **Pindo CLI** you're ready to go.

```bash 
pindo --help
```
```bash
Usage: pindo [OPTIONS] COMMAND [ARGS]...

Pindo CLI
	
A simple Command Line Interface that allows you to authenticate with the Pindo API
	
https://www.pindo.io

Options:
  --debug / --no-debug
  -v, --version         Show the version and exit.
  --help                Show this message and exit.

Commands:
  balance        Get account balance
  org            Organization
  refresh-token  Refresh a Token.
  register       Create a new Pindo account.
  sms            Send a test message
  token          Request a token for using Pindo API.
```
- Send a test message

```bash 
pindo sms --help
```
```bash
Usage: pindo sms [OPTIONS]

  Send a test message

Options:
  --token TEXT   API Token
  --to TEXT      Receiver phone number (+250xxxxxx)
  --text TEXT    Message to send
  --sender TEXT  Sender name
  --help         Show this message and exit.
```

## API Response Code

| **Code** | **Text** | **Meaning** |
---| --- | ---
201 | sent | Successfully sent
401 | unauthorized | unauthorized access
404 | not found | invalid resource URI
409 | conflict | number is from unsupported country
409 | conflict | number is from unsupported telco
409 | conflict | Wrong phone number format

- An example of a successfully sent SMS.

```json
{
    "count": 1,
    "remaining_balance": 3.11,
	"sms_id":"1058918",
    "self_url": "https://api.pindo.io/v1/sms/1058918",
    "sms_items": [
        {
            "id": 1062502,
            "item_price": 0.01,
            "network": "63510",
            "remaining_balance": 3.11,
            "status": "sent",
            "to": "+250785000000"
        }
    ]
}
```

- Pindo Webhook Event example `POST` methods

```json
{
    "status": "DELIVRD",
    "sms_id": 1058918,
    "modified_at": "24-07-2020, 23:35:32",
    "retries_count": 0
}

```

## SMS API Usage

The `pindo api` needs your Token. You can either pass the token directly to the constructor (see the code below) or via environment variables.

```bash

# cURL

curl -X POST \
https://api.pindo.io/v1/sms/ \
-H 'Accept: */*' \
-H 'Authorization: Bearer your-token' \
-H 'Content-Type: application/json' \
-d '{
"to" : "+250781234567",
"text" : "Hello from Pindo",
"sender" : "Pindo"
}'
```

```python

# python

import requests

token='your-token'
headers = {'Authorization': 'Bearer ' + token}
data = {'to' : '+250781234567', 'text' : 'Hello from Pindo', 'sender' : 'Pindo'}

url = 'https://api.pindo.io/v1/sms/'
response = requests.post(url, json=data, headers=headers)
print(response)
print(response.json())

```

```javascript
// NodeJS

var request = require("request");
data = { to: "+250781234567", text: "Hello from Pindo", sender: "Pindo" };

var options = {
  method: "POST",
  body: data,
  json: true,
  url: "https://api.pindo.io/v1/sms/",
  headers: {
    Authorization: "Bearer your-token"
  }
};

function callback(error, response, body) {
  if (!error && response.statusCode == 200) {
    console.log(body);
  }
}
//call the request

request(options, callback);
```

```java

// Java

OkHttpClient client = new OkHttpClient();

MediaType mediaType = MediaType.parse("application/json");
RequestBody body = RequestBody.create(mediaType, "{"to" : "+250781234567", "text" : "Hello from Pindo","sender" : "Pindo"}");
Request request = new Request.Builder()
.url("https://api.pindo.io/v1/sms/")
.post(body)
.addHeader("Content-Type", "application/json")
.addHeader("Authorization", "Bearer your-token")
.build();
Response response = client.newCall(request).execute();
```

```php

// PHP

$request = new HttpRequest();
$request->setUrl('https://api.pindo.io/v1/sms/');
$request->setMethod(HTTP_METH_POST);

$request->setHeaders(array(
'Authorization' => 'Bearer your-token',
'Content-Type' => 'application/json'
));

$request->setBody('{
"to" : "+250781234567",
"text" : "Hello from Pindo",
"sender" : "Pindo"
}');

try {
$response = $request->send();

echo $response->getBody();
} catch (HttpException $ex) {
echo $ex;
}

// cURL

$curl = curl_init();

curl_setopt_array($curl, array(
  CURLOPT_URL => "https://api.pindo.io/v1/sms/",
  CURLOPT_RETURNTRANSFER => true,
  CURLOPT_ENCODING => "",
  CURLOPT_MAXREDIRS => 10,
  CURLOPT_TIMEOUT => 0,
  CURLOPT_FOLLOWLOCATION => true,
  CURLOPT_HTTP_VERSION => CURL_HTTP_VERSION_1_1,
  CURLOPT_CUSTOMREQUEST => "POST",
  CURLOPT_POSTFIELDS =>"{\n\t\"to\" : \"+250781234567\",\n\t\"text\" : \"Test SMS.\",\n\t\"sender\" : \"Pindo\"\n}",
  CURLOPT_HTTPHEADER => array(
    "Authorization: Bearer token",
    "Content-Type: application/json"
  ),
));

$response = curl_exec($curl);

curl_close($curl);
echo $response;
```

```Go

// GO

package main

import (
"fmt"
"strings"
"net/http"
"io/ioutil"
)

func main() {

url := "https://api.pindo.io/v1/sms/"

payload := strings.NewReader("{"to" : "+250781234567", "text" : "Hello from Pindo","sender" : "Pindo"}")

req, _ := http.NewRequest("POST", url, payload)

req.Header.Add("Content-Type", "application/json")
req.Header.Add("Authorization", "Bearer your-token")

res, _ := http.DefaultClient.Do(req)

defer res.Body.Close()
body, _ := ioutil.ReadAll(res.Body)

fmt.Println(res)
fmt.Println(string(body))

}
```

```csharp

// C#

var client = new RestClient("https://api.pindo.io/v1/sms/");
var request = new RestRequest(Method.POST);
request.AddHeader("Authorization", "Bearer your-token");
request.AddHeader("Content-Type", "application/json");
request.AddParameter("undefined", "{\n\t\"to\" : \"+250781234567\", \n\t\"text\" : \"Hello from Pindo\",\n\t\"sender\" : \"Pindo\"\n}", ParameterType.RequestBody);
IRestResponse response = client.Execute(request);
```

```ruby
# ruby

require 'net/http'
require 'json'
require 'uri'

data = { to: '+250781234567', text: 'Hello from Pindo', sender: 'Pindo' };

uri = URI('https://api.pindo.io/v1/sms/')
http = Net::HTTP.new(uri.host, uri.port)
req = Net::HTTP::Post.new(uri)
req['Authorization'] = 'Bearer your-token'
req['Content-Type'] = 'application/json'
req.body = data.to_json
http.request(req)

```

```dart

// Dart

import 'dart:convert';
import 'package:http/http.dart' as http;

Future main() async {
  String url = 'https://api.pindo.io/v1/sms/';
  Map<String, String> data = {
    'to': '+250781234567',
    'text': 'Hello from Pindo',
    'sender': 'Pindo'
  };

  Map<String, String> headers = {
    'Authorization': 'Bearer your-token',
    'Content-Type': 'application/json'
  };

  http.Response response = await http.post(
    url,
    body: jsonEncode(data),
    headers: headers,
  );
  print(response.statusCode);
  print(jsonDecode(response.body));
}

```

## Verify API Usage

`PindoVerfiy` API lets you send a PIN to a user's phone and validate that they received it. PindoVerfiy can be used for a number of authentication and anti-fraud purposes, such as 2-factor authentication, password-less sign-in, and validating users’ phone numbers.

- An example of a successfully generated PIN.

```json
{
    "message": "success",
    "network": "63510",
    "remaining_balance": 487.49,
    "request_id": 4
}
```

- An example of a successfully verified PIN.

```
{
    "message": "success",
    "remaining_balance": 487.49,
    "request_id": 4
}
```


- Generate a PIN.

```python

# python

import requests
import json

url = "https://api.pindo.io/v1/verify"

payload = json.dumps({
  "brand": "Pindo",
  "number": "+250781234567"
})
headers = {
  'Authorization': 'Bearer your-token',
  'Content-Type': 'application/json'
}

response = requests.request("POST", url, headers=headers, data=payload)

print(response.json())
```


```javascript

// NodeJS

var request = require('request');
var options = {
  'method': 'POST',
  'url': 'https://api.pindo.io/v1/verify',
  'headers': {
    'Authorization': 'Bearer your-token',
    'Content-Type': 'application/json'
  },
  body: JSON.stringify({
    "brand": "Pindo",
    "number": "+250781234567"
  })

};
request(options, function (error, response) {
  if (error) throw new Error(error);
  console.log(response.body);
});
```


```java

// Java

OkHttpClient client = new OkHttpClient().newBuilder()
  .build();
MediaType mediaType = MediaType.parse("application/json");
RequestBody body = RequestBody.create(mediaType, "{\n    \"brand\":\"Pindo\",\n    \"number\":\"+250781234567\"\n}\n");
Request request = new Request.Builder()
  .url("https://api.pindo.io/v1/verify")
  .method("POST", body)
  .addHeader("Authorization", "Bearer your-token")
  .addHeader("Content-Type", "application/json")
  .build();
Response response = client.newCall(request).execute();

```

```php

// PHP

<?php

$curl = curl_init();

curl_setopt_array($curl, array(
  CURLOPT_URL => 'https://api.pindo.io/v1/verify',
  CURLOPT_RETURNTRANSFER => true,
  CURLOPT_ENCODING => '',
  CURLOPT_MAXREDIRS => 10,
  CURLOPT_TIMEOUT => 0,
  CURLOPT_FOLLOWLOCATION => true,
  CURLOPT_HTTP_VERSION => CURL_HTTP_VERSION_1_1,
  CURLOPT_CUSTOMREQUEST => 'POST',
  CURLOPT_POSTFIELDS =>'{
    "brand":"Pindo",
    "number":"+250781234567"
}
',
  CURLOPT_HTTPHEADER => array(
    'Authorization: Bearer your-token',
    'Content-Type: application/json'
  ),
));

$response = curl_exec($curl);

curl_close($curl);
echo $response;

```


- Verify a PIN

```javascript

// NodeJS

var request = require('request');
var options = {
  'method': 'POST',
  'url': 'https://api.pindo.io/v1/verify/check',
  'headers': {
    'Authorization': 'Bearer your-token',
    'Content-Type': 'application/json'
  },
  body: JSON.stringify({
    "code": "752623",
    "request_id": 4
  })

};
request(options, function (error, response) {
  if (error) throw new Error(error);
  console.log(response.body);
});

```

- Check PIN status

```javascript

// NodeJS 

var request = require('request');
var options = {
  'method': 'GET',
  'url': 'https://api.pindo.io/v1/verify/status/:request_id',
  'headers': {
    'Authorization': 'Bearer your-token'
  }
};
request(options, function (error, response) {
  if (error) throw new Error(error);
  console.log(response.body);
});

```

- Cancel a PIN

```javascript

// NodeJS

var request = require('request');
var options = {
  'method': 'PUT',
  'url': 'https://api.pindo.io/v1/verify/cancel/:request_id',
  'headers': {
    'Authorization': 'Bearer your-token'
  }
};
request(options, function (error, response) {
  if (error) throw new Error(error);
  console.log(response.body);
});
```


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.2.1 (2022-01-19)
### Fix
* **readme:** Docs ([`5a56a39`](https://github.com/pindoio/pindo-cli/commit/5a56a39d83b0102f8580e004a94dbb03a4d61543))
* **test_cli.py:** Improve test on wrong password and username ([`fbf1c43`](https://github.com/pindoio/pindo-cli/commit/fbf1c43390931bab00bc5d003fa63fa277583cdb))
* **test_cli.py:** Improve test on wrong password and username ([`2ccf818`](https://github.com/pindoio/pindo-cli/commit/2ccf818c0610ebca7883b09e0a74706a8512a5fc))
* **test_cli.py:** Improve test on wrong password and username ([`5aac342`](https://github.com/pindoio/pindo-cli/commit/5aac342b091ab28661fd04bf02471bd9e6995ae6))
* **login component and docs:** Fix token request by supporting wrong password and username and Docs ([`d2a9603`](https://github.com/pindoio/pindo-cli/commit/d2a9603de85341b47c6e125300f663ece8ba4121))

### Documentation
* **readme:** Fix typo ([`4cbd5cc`](https://github.com/pindoio/pindo-cli/commit/4cbd5cc674123bce220b4f4157f308207a979926))

## v0.2.0 (2021-07-23)
### Feature
* **recover-password:** Upgraded pindo-cli to help users recover their password ([`a58ebcc`](https://github.com/pindoio/pindo-cli/commit/a58ebcc3b68fcaa9f1d86d292147edaeaef54b1f))

### Documentation
* Fix docs ([`fd30316`](https://github.com/pindoio/pindo-cli/commit/fd303161848f5218338b2e3411ea467e88a18b3e))
* Documentation only changes ([`1c0a6b0`](https://github.com/pindoio/pindo-cli/commit/1c0a6b0202f48a89dec43c1740bf895731dc3738))
* Documentation only changes ([`af1e583`](https://github.com/pindoio/pindo-cli/commit/af1e583b838e32d7a2629fe7b6690d6230506ab2))
* Documentation only changes ([`597c811`](https://github.com/pindoio/pindo-cli/commit/597c811e1261257d34bb7f516fe0133006b36b7d))
* Documentation only changes ([`0ef1f1a`](https://github.com/pindoio/pindo-cli/commit/0ef1f1a342cec62aeedb091c17d4287691dca56b))

## v0.1.9 (2021-05-19)
### Fix
* Remove windows settings ([`4610ffb`](https://github.com/pindoio/pindo-cli/commit/4610ffba0c9b1ee43265ad4e4ee17c95d5181f83))
* Linting in release.yaml ([`4a20101`](https://github.com/pindoio/pindo-cli/commit/4a20101fc3904536d8e3e1a3d66ac85b53dc2c6f))
* Bug in CI.yaml ([`02a2419`](https://github.com/pindoio/pindo-cli/commit/02a2419907a72e1c919dac7e5225f3c0b8168a7b))
* Bugs in CI action ([`5b4315a`](https://github.com/pindoio/pindo-cli/commit/5b4315a51ad20062180eb6785dac80f614b20593))
* Fix verision in pyproject.toml ([`029dc04`](https://github.com/pindoio/pindo-cli/commit/029dc0429079254121f65d787d4b7cd1381d5f27))


tests/test_cli.py
import pytest
from click.testing import CliRunner
from pkg_resources import get_distribution

import pindo_cli.cli as pindo


@pytest.fixture(scope="module")
def runner():
    return CliRunner()


def test_pindo_cli(runner):
    result = runner.invoke(pindo.cli, ['--help'])
    assert result.exit_code == 0
    assert 'Pindo CLI' in result.output


def test_command_token(runner):
    result = runner.invoke(pindo.token, ['-u', 'dgd', '-p', 'oUhdo'])
    assert result.exit_code == 0
    assert result.output == 'Wrong username or password\n'


def test_command_refresh_token(runner):
    result = runner.invoke(pindo.refresh_token, ['-u', 'abc', '-p', 'cbd'])
    assert result.exit_code == 1
    assert result.output is ''


def test_command_register(runner):
    result = runner.invoke(
        pindo.register,
        ['-u', 'ken', '-e', 'remy@bar.io', '-p', 'bar'])
    assert result.exit_code == 0
    assert 'username already exists in the database' in result.output


def test_command_sms(runner):
    result = runner.invoke(
        pindo.sms, [
            '--token', 'oeiroeoeioeoiroe',
            '--to', '+250785383100',
            '--text', 'Hello Pindo',
            '--sender', 'Pindo'
        ])
    assert result.exit_code == 0
    assert 'message' in result.output


def test_command_balance(runner):
    result = runner.invoke(
        pindo.balance, [
            '--token', 'oeiroeoeioeoiroe',
        ])
    assert result.exit_code == 1
    assert result.output is ''


def test_pindo_cli_version_flag(runner):
    result = runner.invoke(pindo.cli, ['--version', "-v"])
    assert result.exit_code == 0
    assert get_distribution('pindo-cli').version in result.output


def test_command_org(runner):
    result = runner.invoke(
        pindo.org, [
            '--token', 'oeiroeoeioeoiroe',
            '--name', 'Pindo',
            '--webhook_url', 'https://pindo.io/webhook',
            '--retries_count', 10
        ])
    assert result.exit_code == 0
    assert 'message' in result.output


def test_command_forget_password(runner):
    result = runner.invoke(
        pindo.forget_password, [
            '--email', 'luinmeme@gmail.com'
        ])
    assert result.exit_code == 0
    assert 'Wrong email' in result.output


pindo_cli/tests/test_cli.py
import pytest
from click.testing import CliRunner
from pkg_resources import get_distribution

import pindo_cli


@pytest.fixture(scope="module")
def runner():
    return CliRunner()


def test_pindo_cli(runner):
    result = runner.invoke(pindo_cli.cli, ['--help'])
    assert result.exit_code == 0
    assert 'Pindo CLI' in result.output


def test_command_token(runner):
    result = runner.invoke(pindo_cli.token, ['-u', 'ken', '-p', 'bar'])
    assert result.exit_code == 0
    assert 'message' in result.output


def test_command_refresh_token(runner):
    result = runner.invoke(pindo_cli.refresh_token, ['-u', 'ken', '-p', 'bar'])
    assert result.exit_code == 0
    assert 'message' in result.output


def test_command_register(runner):
    result = runner.invoke(
        pindo_cli.register,
        ['-u', 'ken', '-e', 'remy@pindo.io', '-p', 'bar'])
    assert result.exit_code == 0
    assert 'message' in result.output


def test_command_sms(runner):
    result = runner.invoke(
        pindo_cli.sms, [
            '--token', 'oeiroeoeioeoiroe',
            '--to', '+250785383100',
            '--text', 'Hello Pindo',
            '--sender', 'Pindo'
        ])
    assert result.exit_code == 0
    assert 'message' in result.output


def test_command_(runner):
    result = runner.invoke(
        pindo_cli.balance, [
            '--token', 'oeiroeoeioeoiroe',
        ])
    assert result.exit_code == 0
    assert 'message' in result.output


def test_pindo_cli_version_flag(runner):
    result = runner.invoke(pindo_cli.cli, ['--version', "-v"])
    assert result.exit_code == 0
    assert get_distribution('pindo-cli').version in result.output


pindo_cli/http.py
import requests
from requests.auth import HTTPBasicAuth


class PindoClientException(Exception):
    pass


class Config:
    BASE_URL = 'https://api.pindo.io'


class Token(Config):
    """
    Request a Pindo Token
    """
    def __init__(self, username, password):
        self.username = username
        self.password = password
        self.url = '{}/users/token'.format(Config.BASE_URL)

    def __str__(self):
        # request
        r = requests.get(
            self.url, auth=HTTPBasicAuth(self.username, self.password))

        return '{}'.format(r.json()['token'])


class RefreshToken(Config):
    """
    Refresh a Pindo token
    """
    def __init__(self, username, password):
        self.username = username
        self.password = password
        self.url = '{}/users/refresh/token'.format(Config.BASE_URL)

    def __str__(self):
        # request
        r = requests.get(
            self.url, auth=HTTPBasicAuth(self.username, self.password))

        return '{}'.format(r.json()['token'])


class Register(Config):
    """
    Register a new account
    """
    def __init__(self, username, email, password):
        self.username = username
        self.email = email
        self.password = password
        self.url = '{}/users/register'.format(Config.BASE_URL)

    def __str__(self):
        payload = {
            'username': self.username,
            'email': self.email,
            'password': self.password
        }
        r = requests.post(self.url, json=payload)

        return '{}'.format(r.json()['message'])


class ForgetPassword(Config):
    """
    Forget Password
    """
    def __init__(self, email):
        self.email = email
        self.url = '{}/users/forgot'.format(Config.BASE_URL)

    def resp(self):
        payload = {
            'email': self.email
        }
        r = requests.post(self.url, json=payload)

        return r


class RecoverPassword(Config):
    """
    Recover Password
    """
    def __init__(self, token, password, confirm_password):
        self.token = token
        self.password = password
        self.confirm_password = confirm_password
        self.url = '{}/users/recovery/{}'.format(Config.BASE_URL, self.token)

    def __str__(self):
        payload = {
            'password': self.password,
            'confirm_password': self.confirm_password
        }
        r = requests.put(self.url, json=payload)
             
        return '{}'.format(r.json()['message'])

    
class SMS(Config):
    """
    Send a test Message
    """

    def __init__(self, token, to, text, sender):
        self.token = token
        self.to = to
        self.text = text
        self.sender = sender
        self.url = '{}/v1/sms/'.format(Config.BASE_URL)

    def send(self):
        payload = {
            'to': self.to,
            'text': self.text,
            'sender': self.sender
        }

        headers = {'Authorization': 'Bearer ' + self.token}

        r = requests.post(
            self.url,
            headers=headers,
            json=payload
        )

        return r.json()


class Balance(Config):
    """
    Get user's wallet
    """
    def __init__(self, token):
        self.token = token

    def __str__(self):
        headers = {'Authorization': 'Bearer ' + self.token}
        wallet_url = '{}/wallets/self'.format(Config.BASE_URL)
        r = requests.get(wallet_url, headers=headers)

        return '{}'.format(r.json()['amount'])


class Organization(Config):
    """
    Organization setup
    """
    def __init__(self, token, name, webhook_url, retries_count):
        self.token = token
        self.name = name
        self.webhook_url = webhook_url
        self.retries_count = retries_count

    def __str__(self):
        payload = {
            'name': self.name,
            'webhook_url': self.webhook_url,
            'sms_retries': self.retries_count
        }
        headers = {'Authorization': 'Bearer ' + self.token}
        org_url = '{}/orgs/self'.format(Config.BASE_URL)
        r = requests.put(
            org_url,
            headers=headers,
            json=payload
        )

        return '{}'.format(r.json())


pindo_cli/__init__.py
import click
import click_spinner

from pindo_cli.http import Token, RefreshToken, Register, SMS, Balance, Organization


@click.group()
@click.option('--debug/--no-debug', default=False)
@click.version_option(None, "--version", "-v")
def cli(debug):
    """
    Pindo CLI
    A simple Command Line Interface that allow you to create an account
    and request a token for using Pindo API
    """


@cli.command()  # @cli, not @click!
@click.option(
    '--username', '-u', prompt=True, help='Please enter your username')
@click.option(
    '--password', '-p',
    prompt=True, hide_input=True, help='Please enter your password')
def token(username, password):
    """
    Request a token for using Pindo API.
    """
    click.echo('token...')
    with click_spinner.spinner():
        click.echo(Token(username, password))


@cli.command()  # @cli, not @click!
@click.option(
    '--username', '-u', prompt=True, help='Please enter your username')
@click.option(
    '--password', '-p',
    prompt=True, hide_input=True, help='Please enter your password')
def refresh_token(username, password):
    """
    Refresh a Token.
    """
    click.echo('token...')
    with click_spinner.spinner():
        click.echo(RefreshToken(username, password))


@cli.command()
@click.option(
    '--username', '-u', prompt=True, help='Please enter your username')
@click.option(
    '--email', '-e', prompt=True, help='Please enter your email')
@click.option(
    '--password', '-p', prompt=True, hide_input=True, confirmation_prompt=True,
    help='Please enter your password')
def register(username, email, password):
    """
    Create a new Pindo account.
    """
    click.echo(Register(username, email, password))


@cli.command()
@click.option('--token', prompt=True, help='API Token')
@click.option(
    '--to', prompt=True, help='Receiver phone number (+250xxxxxx)',
    default='+250785383100')
@click.option(
    '--text', prompt=True, help='Message to send', default='Hello Pindo')
@click.option('--sender', prompt=True, default='Pindo', help='Sender name')
def sms(token, to, text, sender):
    """
    Send a test message
    """
    click.echo(SMS(token, to, text, sender))


@cli.command()
@click.option('--token', prompt=True, help='API Token')
def balance(token):
    """
    Get account balance
    """
    click.echo(Balance(token))


@cli.command()
@click.option('--token', prompt=True, help='API Token')
@click.option('--name', prompt=True, help='Organization name')
@click.option('--webhook_url', prompt=True, help='Webhook URL. Eg (https://pindo.io/dlr)')
@click.option('--retries_count', prompt=True, type=(int), help='SMS retries count settings')
def org(token, name, webhook_url, retries_count):
    """
    Organization
    """
    click.echo(Organization(token, name, webhook_url, retries_count))


pindo_cli/cli.py
import click
import click_spinner
from rich import print
from rich.panel import Panel
from rich.syntax import Syntax
from rich.console import Console
from rich.pretty import Pretty

from pindo_cli.http import Token, RefreshToken, Register, SMS, Balance, Organization, ForgetPassword, RecoverPassword


console = Console()

@click.group()
@click.option('--debug/--no-debug', default=False)
@click.version_option(None, "--version", "-v")
def cli(debug):
    """
    Pindo CLI

    A simple Command Line Interface that allows you to authenticate with 
    the Pindo API
    
    https://www.pindo.io/
    """


@cli.command()  # @cli, not @click!
@click.option(
    '--username', '-u', prompt=True, required=True, help='Please enter your username')
@click.option(
    '--password', '-p',
    prompt=True, hide_input=True, required=True, help='Please enter your password')
def token(username, password):
    """
    Request a token for using Pindo API.
    """
    with console.status(""):
        try:
            token = Panel(str(Token(username, password)), title="API Token", expand=False)
            console.print(token)
        except KeyError:
            console.print("Wrong username or password")


@cli.command()  # @cli, not @click!
@click.option(
    '--username', '-u', prompt=True, required=True, help='Please enter your username')
@click.option(
    '--password', '-p',
    prompt=True, hide_input=True, required=True, help='Please enter your password')
def refresh_token(username, password):
    """
    Refresh a Token.
    """
    with console.status(""):
        token = Panel(str(RefreshToken(username, password)), title="New API Token", expand=False)
        console.print(token, style="bold green")


@cli.command()
@click.option(
    '--username', '-u', prompt=True, required=True, help='Please enter your username')
@click.option(
    '--email', '-e', prompt=True, required=True, help='Please enter your email')
@click.option(
    '--password', '-p', prompt=True, required=True, hide_input=True, confirmation_prompt=True,
    help='Please enter your password')
def register(username, email, password):
    """
    Create a new Pindo account.
    """
    with console.status(""):
        console.print(str(Register(username, email, password)))


@cli.command()
@click.option('--email', '-e', prompt=True, required=True, help='Please enter your email')
@click.pass_context
def forget_password(ctx, email):
    """
    Forget Password
    """
    with console.status(""):
        resp = ForgetPassword(email).resp()
        if resp.status_code == 200:
            print('A recover token has been sent to your email. Run the command [bold green]pindo recover-password[/bold green] to recover your password account.')
            syntax = Syntax('run: pindo recover-password', "bash", theme="monokai", line_numbers=False)
            panel = Panel(syntax, title="Command", expand=False)
            console.print(panel)
        else:
            panel = Panel('The email provided does not exist.', title="Wrong email", expand=False)
            console.print(panel)


@cli.command(hidden=True)
@click.option('--token', '-t', prompt=True, required=True, help='Enter the recover token sent to your email.')
@click.option('--password', prompt=True, hide_input=True, required=True, help='Enter your new password')
@click.option('--confirm_password', prompt=True, hide_input=True, required=True, help='Confirm your password')
def recover_password(token, password, confirm_password):
    """
    Recover Password
    """
    with console.status(""):
        console.print(str(RecoverPassword(token, password, confirm_password)))


@cli.command()
@click.option('--token', prompt=True, required=True, help='API Token')
@click.option(
    '--to', prompt=True, required=True, help='Receiver phone number (+250xxxxxx)',
    default='+250785383100')
@click.option(
    '--text', prompt=True, required=True, help='Message to send', default='Hello Pindo')
@click.option('--sender', prompt=True, required=True, default='Pindo', help='Sender name')
def sms(token, to, text, sender):
    """
    Send a test message
    """
    with console.status(""):
        sms = SMS(token, to, text, sender).send()
        pretty = Pretty(sms, expand_all=True)
        console.print(Panel(pretty, title="SMS", expand=False))


@cli.command()
@click.option('--token', prompt=True, required=True, help='API Token')
def balance(token): 
    """
    Get account balance
    """
    with console.status(""):
        balance = "USD " + str(Balance(token))
        panel =Panel(balance, title="Balance", expand=False)
        console.print(panel)


@cli.command()
@click.option('--token', prompt=True, required=True, help='API Token')
@click.option('--name', prompt=True, required=True, help='Organization name')
@click.option('--webhook_url', prompt=True, required=True, help='Webhook URL. Eg (https://pindo.io/dlr)')
@click.option('--retries_count', prompt=True, type=(int), required=True, help='SMS retries count settings')
def org(token, name, webhook_url, retries_count):
    """
    Update Organization Settings
    """
    console.print(str(Organization(token, name, webhook_url, retries_count)))
